pkg/controllers/utils.go

```diff
@@ -1,91 +1,213 @@
 	}
 }
 
+// PortToServicePort converts an internal model port to a k8s service port
 func PortToServicePort(port model.Port) v1.ServicePort {
 	return v1.ServicePort{
 		Name:       port.Name,
-		Protocol:   stringToProtocol(port.Protocol),
+		Protocol:   v1.Protocol(port.Protocol),
 		Port:       port.Port,
 		TargetPort: intstr.Parse(port.TargetPort),
 	}
 }
 
+// PortToServiceImportPort converts an internal model port to a service import port
+func PortToServiceImportPort(port model.Port) v1alpha1.ServicePort {
+	return v1alpha1.ServicePort{
+		Name:     port.Name,
+		Protocol: v1.Protocol(port.Protocol),
+		Port:     port.Port,
+	}
+}
+
+// PortToEndpointPort converts an internal model port to a k8s endpoint port
 func PortToEndpointPort(port model.Port) discovery.EndpointPort {
-	protocol := stringToProtocol(port.Protocol)
+	protocol := v1.Protocol(port.Protocol)
 	return discovery.EndpointPort{
 		Name:     &port.Name,
 		Protocol: &protocol,
 		Port:     &port.Port,
 	}
 }
 
-func protocolToString(protocol v1.Protocol) string {
-	switch protocol {
-	case v1.ProtocolTCP:
-		return model.TCPProtocol
-	case v1.ProtocolUDP:
-		return model.UDPProtocol
-	case v1.ProtocolSCTP:
-		return model.SCTPProtocol
-	default:
-		return ""
+// ExtractServicePorts extracts all unique service ports from a slice of endpoints
+func ExtractServicePorts(endpoints []*model.Endpoint) (servicePorts []*model.Port) {
+	uniquePorts := make(map[string]model.Port)
+	for _, ep := range endpoints {
+		uniquePorts[ep.ServicePort.GetID()] = ep.ServicePort
+	}
+	for _, servicePort := range uniquePorts {
+		portRef := servicePort
+		servicePorts = append(servicePorts, &portRef)
 	}
+	return servicePorts
 }
 
-func stringToProtocol(protocol string) v1.Protocol {
-	switch protocol {
-	case model.TCPProtocol:
-		return v1.ProtocolTCP
-	case model.UDPProtocol:
-		return v1.ProtocolUDP
-	case model.SCTPProtocol:
-		return v1.ProtocolSCTP
-	default:
-		return ""
+// ExtractEndpointPorts extracts all unique endpoint ports from a slice of endpoints
+func ExtractEndpointPorts(endpoints []*model.Endpoint) (endpointPorts []*model.Port) {
+	uniquePorts := make(map[string]model.Port)
+	for _, ep := range endpoints {
+		uniquePorts[ep.EndpointPort.GetID()] = ep.EndpointPort
+	}
+	for _, endpointPort := range uniquePorts {
+		portRef := endpointPort
+		endpointPorts = append(endpointPorts, &portRef)
 	}
+	return endpointPorts
 }
 
-func EndpointPortsAreEqualIgnoreOrder(a, b []discovery.EndpointPort) (equal bool) {
+func PortsEqualIgnoreOrder(a, b []*model.Port) (equal bool) {
 	if len(a) != len(b) {
 		return false
 	}
 
-	for _, aPort := range a {
-		match := false
-		for _, bPort := range b {
-			if reflect.DeepEqual(aPort, bPort) {
-				match = true
-				break
-			}
-		}
-		if !match {
+	aMap := make(map[string]*model.Port)
+	for _, portA := range a {
+		aMap[portA.GetID()] = portA
+	}
+
+	for _, portB := range b {
+		if !portB.Equals(aMap[portB.GetID()]) {
```

Comment:
should we add check if `portB.GetID()` key is not found in the `aMap`

Reply:
Good catch